libsnark/zk_proof_systems/plonk/utils.tcc

```diff
@@ -382,6 +382,81 @@ bool plonk_are_valid_constants_k1_k2(const FieldT &k1, const FieldT &k2)
+// determined by the i-th gate of the arithmetic circuit, which can be
+// one of the following: addition, multiplication, multiplication by
+// constant, public input. In particular, when the i-th gate is a
+// public input, the i-th components of the selector vectors are:
+//
+// (q_L[i], q_R[i], q_O[i], q_M[i], q_C[i]) = (1, 0, 0, 0, 0)
+//
+// Therefore the top N rows of the initialized gates matrix will have
+// the above form. See also Section 6 [GWC19].
+//
+// \attention The convention for selector vector values corresponding
+// to public inputs (PI) used in this function follows [GWC19] (as
+// shown above) and is different from the one used in the example
+// circuit for BLS12-381 (class example) where q_R[i]=1 rather than
+// q_L[i]=1 i.e.
+//
+// (q_L[i], q_R[i], q_O[i], q_M[i], q_C[i]) = (0, 1, 0, 0, 0)
+template<typename ppT>
+std::vector<std::vector<libff::Fr<ppT>>> plonk_prepare_gates_matrix(
+    const size_t &num_public_inputs)
+{
+    using FieldT = libff::Fr<ppT>;
+    std::vector<std::vector<FieldT>> gates_matrix_init;
+    const std::vector<FieldT> PI_selector_vector{1, 0, 0, 0, 0};
+    for (size_t i = 0; i < num_public_inputs; ++i) {
+        gates_matrix_init.push_back(PI_selector_vector);
+    }
+    return gates_matrix_init;
+}
+
+// Extract the values corresponing to the public inputs from the
+// witness using the respective wire indices passed as input. Those
+// values are passed on to the verifier together with the proof.
+template<typename ppT>
+std::vector<libff::Fr<ppT>> plonk_public_input_values_from_indices(
+    const std::vector<libff::Fr<ppT>> &witness,
+    const std::vector<size_t> &PI_wire_indices)
+{
+    assert(PI_wire_indices.size() <= witness.size());
+
+    using FieldT = libff::Fr<ppT>;
+    std::vector<FieldT> PI_value_list;
+    for (size_t i = 0; i < PI_wire_indices.size(); i++) {
+        assert(PI_wire_indices[i] < witness.size());
+        FieldT PI_value = witness[PI_wire_indices[i]];
+        PI_value_list.push_back(PI_value);
+    }
+    return PI_value_list;
+}
+
+// A wrapper for plonk_public_input_values_from_indices. Extracts the
+// values corresponing to the public inputs from the witness, assuming
+// that they are in the first num_public_inputs positions. In other
+// words extracts the values of the public inputs from the *length* of
+// the public input vector.
+template<typename ppT>
+std::vector<libff::Fr<ppT>> plonk_public_input_values_from_length(
+    const std::vector<libff::Fr<ppT>> &witness, const size_t &num_public_inputs)
+{
+    assert(num_public_inputs <= witness.size());
+
+    using FieldT = libff::Fr<ppT>;
+    std::vector<size_t> PI_wire_indices;
+    // store the indices of the PIs
+    for (size_t i = 0; i < num_public_inputs; ++i) {
+        PI_wire_indices.push_back(i);
+    }
+    std::vector<FieldT> PI_value_list =
+        plonk_public_input_values_from_indices<ppT>(witness, PI_wire_indices);
+    return PI_value_list;
```

Comment:
I see the point about re-using the more complex code, but in this case it's so trivial to extract the first `n` entries of a vector.  Something like:
```suggestion
    assert(num_public_inputs <= witness.size());
    return std::vector(witness.begin(), witness.begin() + num_public_inputs);
```

Reply:
Much cleaner indeed. Fixed.